added some methods needed by the mobile wallets

Skvuddudutt opened the pull request:
title

Senyoret1 commented on mobile/api.go:
```diff
@@ -2,19 +2,63 @@ package mobile
-func GetAddresses(seed string, amount int) (string, error) {
-	addresses := liteclient.GenerateAddresses(seed, amount)
+// Seed should be the mnemonic word list
+func GetAddresses(seed string, num int) (string, error) {
```
Most of the code of this function could be deleted simply by calling `liteclient.GenerateAddresses`, which would have the advantage of avoiding repeating the code and that there are already some tests for `liteclient.GenerateAddresses`. The web wallet uses that function and works correctly with the seeds made with the desktop wallet. Is there any particular reason for repeating the code here?

Another thing is that, as returning multiple values is something not very common in other languages, until now almost all the functions on this repository call `panic()` in case of error instead of returning the error, to throw exceptions (this function maintained the Go format because it is old code that had not been touched and was not in use). In fact, in the `PrepareTransaction` function below it would work like this, because `liteclient.PrepareTransaction` calls `panic`. If there is no particular reason for returning the error, it might be good to maintain the standard used so far.

Skvuddudutt replied:
I remember there was *some* reason I did the GenerateAddresses again but I cant remember why :-P I'll take a look at it and change it or find the reason.

I think I found that the gomobile java-generator automatically wraps the "val,error" format into throwing an exception if the error is not nil, but I dont really care about it so I'll change it to the panic() way of doing it if you prefer it. I'll update this PR soon.

Skvuddudutt commented:
sorry for taking so long. I've updated my changes in the mobile.go api to call the liteclient instead when creating addresses.

The reason for the old code was that the gomobile wrapper tool doesnt like panic(). When wrapping to Java it automatically creates Java-Exceptions for the go-idiom of returning the tuple (value, error). However if a go function instead relies on calling panic()-recover() the tool cant handle it.

But I fixed it now and added some manual handling of the panic() so the function gets a (value,error) format return.

Senyoret1 commented:
There are two things that may cause problems.

The first is that, at least in the tests I did on desktop, I did not try Go Mobile, putting a `recover()` in the functions of the `mobile` package is not enough. If, for example, you modify the `GetAddresses` function in the `mobile` package, with a code like this, the recover works correctly:

```
byteaddr, err := json.Marshal(addresses)
if err != nil {
	return "", err
} else {
	panic("force panic")
}
```

It works because the panic is in the same package. However, if instead you modify the `GenerateAddresses` function of the `liteclient` package so that it always calls `panic()`, the `recover()` will not work, the execution will stop.

The current approach of using panics was chosen because in Gopherjs the panics are converted into errors, because in Javascript there is no clear concept of "closing the application". However, if it is causing problems to the mobile wallet, maybe it would be more appropriate to change all the methods on `liteclient` to return errors instead of using `panic()`, so the code for the mobile wallet would be very simple, and then make the functions called by the JS code convert those errors into panics. If it seems right, I can make the change. What do you think, @gz-c ?

The second problematic thing is that the `GetAddresses` function of the `mobile` package calls `hex.EncodeToString`. I do not know if I did not take into account something, but that seems to make it difficult to get new addresses. For example, if I call the function with the seed "test" and ask for 2 addresses, I get this:

![r1](https://user-images.githubusercontent.com/34079003/44051158-10312e14-9f07-11e8-951d-37d0a50bb247.png)

I should be able to get the second address by calling the function if I send the `NextSeed` value of the first address as the seed. However, I get this:

![r2](https://user-images.githubusercontent.com/34079003/44051171-16280aa4-9f07-11e8-9c2c-78f470f309bd.png)

To work properly, `hex.EncodeToString` should be outside the `GetAddresses` function and be used only when the seed is a nmothecnic string. However, I do not know if I'm missing something, because I do not know how the function is going to be used in the mobile wallet.

Skvuddudutt commented:
I'm not very experienced with go but I did test the defer->recover by forcing a panic down in liteclient.PrepareTransaction() and it worked fine in turning it into a (value,error) return. But if you are sure there are cases it wont work then I have to change it for sure.

As you say, the options are either to rewrite a bunch of liteclient code, or copy it into two versions one for the mobile api and one for Gopherjs. I'm not a fan of either of these to be honest. Is there no way to reliably recover from a deep nested panic() in go?

As for the other issue, you are right in that it wont be able to take "nextSeed" but the mobile wallet does not use nextSeed, it always iterates "num" times from the base seed. Thats why the method is commented that the "seed" string should be a word seed. This is to be able to only store the base mnemonic seed and send it as-is into the GetAddresses function.

If the mobile wallet in the future needs to be able to use "nextSeed" its easy to make a new mobile-api-only method that can accept that. Alternatively I could just add a flag to the GetAddress-parameters to indicate wether to treat the seed as a hex-encoded string or a raw byte array. I have no strong feelings about that, I just wanted to get a useable api for the mobile client as simply as possible since I'm not really a go expert.

I feel the liteclient is more the domain of you guys, if you two just let me know what you prefer I'll do whats needed.

Skvuddudutt commented:
great! Who can merge this then? @gz-c maybe?

gz-c commented:
Note that some deep panics are absolutely fatal and indicate an issue with the core crypto libraries.

We should add some notes about the difference in error/panic handling done in the mobile and js bindings
